In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

print("Libraries loaded successfully!")

Libraries loaded successfully!


In [2]:
df = pd.read_csv("../data/raw/Online Retail.csv")

In [3]:
df.head()

,index,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,12/1/2010 8:26,2.55,17850.0,United Kingdom
1,1,536365,71053,WHITE METAL LANTERN,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
2,2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,12/1/2010 8:26,2.75,17850.0,United Kingdom
3,3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
4,4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,12/1/2010 8:26,3.39,17850.0,United Kingdom


In [4]:
df.shape


(541909, 9)

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 9 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   index        541909 non-null  int64  
 1   InvoiceNo    541909 non-null  object 
 2   StockCode    541909 non-null  object 
 3   Description  540455 non-null  object 
 4   Quantity     541909 non-null  int64  
 5   InvoiceDate  541909 non-null  object 
 6   UnitPrice    541909 non-null  float64
 7   CustomerID   406829 non-null  float64
 8   Country      541909 non-null  object 
dtypes: float64(2), int64(2), object(5)
memory usage: 37.2+ MB


In [6]:
df.describe()

,index,Quantity,UnitPrice,CustomerID
count,541909.00000,541909.000000,541909.000000,406829.000000
mean,270954.00000,9.552250,4.611114,15287.690570
std,156435.79785,218.081158,96.759853,1713.600303
min,0.00000,-80995.000000,-11062.060000,12346.000000
25%,135477.00000,1.000000,1.250000,13953.000000
50%,270954.00000,3.000000,2.080000,15152.000000
75%,406431.00000,10.000000,4.130000,16791.000000
max,541908.00000,80995.000000,38970.000000,18287.000000


In [7]:
df.isnull().sum()

index               0
InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

In [8]:
print("Missing values:\n", df.isnull().sum())
print("\nDuplicate rows:", df.duplicated().sum())
print("\nCancelled orders (InvoiceNo starts with 'C'):", df['InvoiceNo'].astype(str).str.startswith('C').sum())
print("\nNegative/zero Quantity:", (df['Quantity'] <= 0).sum())
print("\nNegative/zero UnitPrice:", (df['UnitPrice'] <= 0).sum())

Missing values:
 index               0
InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

Duplicate rows: 0

Cancelled orders (InvoiceNo starts with 'C'): 9288

Negative/zero Quantity: 10624

Negative/zero UnitPrice: 2517


In [9]:
df.tail()

,index,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
541904,541904,581587,22613,PACK OF 20 SPACEBOY NAPKINS,12,12/9/2011 12:50,0.85,12680.0,France
541905,541905,581587,22899,CHILDREN'S APRON DOLLY GIRL,6,12/9/2011 12:50,2.10,12680.0,France
541906,541906,581587,23254,CHILDRENS CUTLERY DOLLY GIRL,4,12/9/2011 12:50,4.15,12680.0,France
541907,541907,581587,23255,CHILDRENS CUTLERY CIRCUS PARADE,4,12/9/2011 12:50,4.15,12680.0,France
541908,541908,581587,22138,BAKING SET 9 PIECE RETROSPOT,3,12/9/2011 12:50,4.95,12680.0,France


In [10]:
df[ df['UnitPrice'] < 0 ]

,index,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
299983,299983,A563186,B,Adjust bad debt,1,8/12/2011 14:51,-11062.06,NaN,United Kingdom
299984,299984,A563187,B,Adjust bad debt,1,8/12/2011 14:52,-11062.06,NaN,United Kingdom


In [11]:
df[ df['Quantity'] < 0 ]

,index,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
141,141,C536379,D,Discount,-1,12/1/2010 9:41,27.50,14527.0,United Kingdom
154,154,C536383,35004C,SET OF 3 COLOURED FLYING DUCKS,-1,12/1/2010 9:49,4.65,15311.0,United Kingdom
235,235,C536391,22556,PLASTERS IN TIN CIRCUS PARADE,-12,12/1/2010 10:24,1.65,17548.0,United Kingdom
236,236,C536391,21984,PACK OF 12 PINK PAISLEY TISSUES,-24,12/1/2010 10:24,0.29,17548.0,United Kingdom
237,237,C536391,21983,PACK OF 12 BLUE PAISLEY TISSUES,-24,12/1/2010 10:24,0.29,17548.0,United Kingdom
...,...,...,...,...,...,...,...,...,...
540449,540449,C581490,23144,ZINC T-LIGHT HOLDER STARS SMALL,-11,12/9/2011 9:57,0.83,14397.0,United Kingdom
541541,541541,C581499,M,Manual,-1,12/9/2011 10:28,224.69,15498.0,United Kingdom
541715,541715,C581568,21258,VICTORIAN SEWING BOX LARGE,-5,12/9/2011 11:57,10.95,15311.0,United Kingdom
541716,541716,C581569,84978,HANGING HEART JAR T-LIGHT HOLDER,-1,12/9/2011 11:58,1.25,17315.0,United Kingdom


In [12]:
df[~df['StockCode'].str[0].str.isdigit()]['StockCode'].unique()

array(['POST', 'D', 'C2', 'DOT', 'M', 'BANK CHARGES', 'S', 'AMAZONFEE',
       'DCGS0076', 'DCGS0003', 'gift_0001_40', 'DCGS0070', 'm',
       'gift_0001_50', 'gift_0001_30', 'gift_0001_20', 'DCGS0055',
       'DCGS0072', 'DCGS0074', 'DCGS0069', 'DCGS0057', 'DCGSSBOY',
       'DCGSSGIRL', 'gift_0001_10', 'PADS', 'DCGS0004', 'DCGS0073',
       'DCGS0071', 'DCGS0068', 'DCGS0067', 'DCGS0066P', 'B', 'CRUK'],
      dtype=object)

In [13]:
df[df['StockCode'] == 'DCGS0076']

,index,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
21326,21326,538071,DCGS0076,SUNJAR LED NIGHT NIGHT LIGHT,1,12/9/2010 14:09,16.13,NaN,United Kingdom
39313,39313,539631,DCGS0076,SUNJAR LED NIGHT NIGHT LIGHT,2,12/20/2010 15:03,16.13,NaN,United Kingdom


In [14]:
df[df['StockCode'] == 'POST']

,index,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
45,45,536370,POST,POSTAGE,3,12/1/2010 8:45,18.0,12583.0,France
386,386,536403,POST,POSTAGE,1,12/1/2010 11:27,15.0,12791.0,Netherlands
1123,1123,536527,POST,POSTAGE,1,12/1/2010 13:04,18.0,12662.0,Germany
5073,5073,536840,POST,POSTAGE,1,12/2/2010 18:27,18.0,12738.0,Germany
5258,5258,536852,POST,POSTAGE,1,12/3/2010 9:51,18.0,12686.0,France
...,...,...,...,...,...,...,...,...,...
541198,541198,581493,POST,POSTAGE,1,12/9/2011 10:10,15.0,12423.0,Belgium
541216,541216,581494,POST,POSTAGE,2,12/9/2011 10:13,18.0,12518.0,Germany
541730,541730,581570,POST,POSTAGE,1,12/9/2011 11:59,18.0,12662.0,Germany
541767,541767,581574,POST,POSTAGE,2,12/9/2011 12:09,18.0,12526.0,Germany


In [15]:
df[df["StockCode"].isin(["POST", "D", "M"])][
    ["InvoiceNo", "StockCode", "Description",
     "Quantity", "InvoiceDate", "UnitPrice", "CustomerID"]
].head(20)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID
45,536370,POST,POSTAGE,3,12/1/2010 8:45,18.00,12583.0
141,C536379,D,Discount,-1,12/1/2010 9:41,27.50,14527.0
386,536403,POST,POSTAGE,1,12/1/2010 11:27,15.00,12791.0
1123,536527,POST,POSTAGE,1,12/1/2010 13:04,18.00,12662.0
2239,536569,M,Manual,1,12/1/2010 15:35,1.25,16274.0
2250,536569,M,Manual,1,12/1/2010 15:35,18.95,16274.0
5073,536840,POST,POSTAGE,1,12/2/2010 18:27,18.00,12738.0
5258,536852,POST,POSTAGE,1,12/3/2010 9:51,18.00,12686.0
5325,536858,POST,POSTAGE,2,12/3/2010 10:36,40.00,13520.0
5369,536861,POST,POSTAGE,3,12/3/2010 10:44,18.00,12427.0


In [16]:
df[df["StockCode"].str.contains("BANK", case=False, na=False)][
    ["InvoiceNo", "StockCode", "Description",
     "Quantity", "UnitPrice", "CustomerID"]
].head(20)

,InvoiceNo,StockCode,Description,Quantity,UnitPrice,CustomerID
4406,536779,BANK CHARGES,Bank Charges,1,15.00,15823.0
14435,C537572,BANK CHARGES,Bank Charges,-1,95.38,NaN
28992,C538680,BANK CHARGES,Bank Charges,-1,966.92,NaN
62508,541505,BANK CHARGES,Bank Charges,1,15.00,15939.0
64573,C541653,BANK CHARGES,Bank Charges,-1,1050.15,NaN
90558,C544049,BANK CHARGES,Bank Charges,-1,566.37,NaN
96675,C544575,BANK CHARGES,Bank Charges,-1,134.76,NaN
96676,C544576,BANK CHARGES,Bank Charges,-1,149.16,NaN
96715,C544584,BANK CHARGES,Bank Charges,-1,109.84,NaN
112233,C545887,BANK CHARGES,Bank Charges,-1,326.68,NaN


In [17]:
df[df["StockCode"].str.contains("GIFT", case=False, na=False)][
    ["InvoiceNo", "StockCode", "Description",
     "Quantity", "UnitPrice", "CustomerID"]
].head(20)

,InvoiceNo,StockCode,Description,Quantity,UnitPrice,CustomerID
38248,539492,gift_0001_40,Dotcomgiftshop Gift Voucher œ40.00,1,34.04,NaN
42057,539958,gift_0001_50,Dotcomgiftshop Gift Voucher œ50.00,1,42.55,NaN
44725,540238,gift_0001_30,Dotcomgiftshop Gift Voucher œ30.00,1,25.53,NaN
44794,540238,gift_0001_20,Dotcomgiftshop Gift Voucher œ20.00,1,17.02,NaN
55589,540995,gift_0001_20,Dotcomgiftshop Gift Voucher œ20.00,1,16.67,NaN
91278,544089,gift_0001_20,Dotcomgiftshop Gift Voucher œ20.00,1,16.67,NaN
94054,544323,gift_0001_30,Dotcomgiftshop Gift Voucher œ30.00,1,25.00,NaN
95034,544434,gift_0001_30,Dotcomgiftshop Gift Voucher œ30.00,1,25.00,NaN
112442,545895,gift_0001_10,Dotcomgiftshop Gift Voucher œ10.00,1,8.33,NaN
145463,548893,gift_0001_40,Dotcomgiftshop Gift Voucher œ40.00,1,33.33,NaN


In [18]:
df[df['StockCode'].str.startswith('DCGS')]

,index,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
21326,21326,538071,DCGS0076,SUNJAR LED NIGHT NIGHT LIGHT,1,12/9/2010 14:09,16.13,NaN,United Kingdom
24906,24906,538349,DCGS0003,BOXED GLASS ASHTRAY,1,12/10/2010 14:59,2.51,NaN,United Kingdom
36460,36460,539451,DCGS0003,BOXED GLASS ASHTRAY,1,12/17/2010 16:59,2.51,NaN,United Kingdom
39313,39313,539631,DCGS0076,SUNJAR LED NIGHT NIGHT LIGHT,2,12/20/2010 15:03,16.13,NaN,United Kingdom
40052,40052,539718,DCGS0070,CAMOUFLAGE DOG COLLAR,1,12/21/2010 13:06,12.72,NaN,United Kingdom
74825,74825,542529,DCGS0055,NaN,-1,1/28/2011 13:08,0.00,NaN,United Kingdom
74838,74838,542531,DCGS0072,NaN,-1,1/28/2011 13:08,0.00,NaN,United Kingdom
74839,74839,542532,DCGS0074,NaN,-1,1/28/2011 13:09,0.00,NaN,United Kingdom
75053,75053,542541,DCGS0069,OOH LA LA DOGS COLLAR,1,1/28/2011 14:25,15.79,NaN,United Kingdom
75295,75295,542582,DCGS0057,NaN,-6,1/28/2011 15:01,0.00,NaN,United Kingdom


In [19]:
df["Country"].nunique()

38

In [20]:
df["Country"].value_counts()

Country
United Kingdom          495478
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
USA                        291
Hong Kong                  288
Singapore                  229
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United Arab Emirates        68
European Community          61
RSA                         58


In [21]:
df["InvoiceDate"].min()

'1/10/2011 10:04'

In [22]:
df["InvoiceDate"].max()

'9/9/2011 9:52'

In [23]:
df["InvoiceDate"] = pd.to_datetime(df["InvoiceDate"])

In [24]:
df["InvoiceDate"].dtype

dtype('<M8[ns]')

In [25]:
df["InvoiceDate"].max() - df["InvoiceDate"].min()

Timedelta('373 days 04:24:00')

In [26]:
df["CustomerID"].nunique()

4372

In [27]:
df["InvoiceNo"].nunique()

25900

In [28]:
len(df)

541909

In [29]:
df.duplicated().sum()

np.int64(0)

In [30]:
df["CustomerID"].isna().sum()

np.int64(135080)

## Exploration Summary

**Dataset scope**
- Total transactions (rows): 541,909
- Unique customers: 4,372
- Unique invoices: 25,900
- Countries: 38
- Date range: '1/10/2011 10:04' to '9/9/2011 9:52' (373 days)

**Issues found**
1. Missing `CustomerID` in ~135,080 rows — cannot attribute these to a customer
2. Missing `Description` in 1,454 rows
3. `InvoiceDate` stored as object, not datetime
4. Negative `UnitPrice`: 2 rows, "Adjust bad debt" (StockCode B) — accounting entries, not sales
5. Negative `Quantity`: 10,624 rows — customer returns, with valid CustomerIDs
6. 33 non-numeric StockCodes, including fees (POST, DOT, BANK CHARGES, AMAZONFEE), gift cards, and admin codes
7. Duplicates: none

**Planned cleaning actions (notebook 02)**
- Drop rows with missing CustomerID (CLV requires customer attribution)
- Drop bad debt and non-product StockCodes
- Keep returns (negative quantity) so TotalAmount nets out correctly
- Convert InvoiceDate to datetime
- Create TotalAmount = Quantity × UnitPrice